### DISCUSSION TITLE
[Release] 1.4

### DISCUSSION THREAD

**Discussion by oxedions** [bluebanquise/bluebanquise#474] [General] [answered]
I wish to release if possible 1.4 before the end of this year.

My main objectives are described in the associated project https://github.com/bluebanquise/bluebanquise/projects/4

Currently, main concern on my side is the diskless part, but this part should speedup soon once the dev in charge is back online.

Does anyone see other needs for this 1.4 release ?

**Comment by strus38** [comment]
could you put the Singularity role in this 1.4?
Thanks

**Reply by btravouillon** [reply]
This role is already in master so it will be available in 1.4.

**Comment by oxedions** [comment]
Hello all.

After investigating, so far for 1.4, I see in the "incoming pipe" :

* one PR regarding Prometheus in MI
* one PR regarding PXE and bootset
* maybe one doc PR on Prometheus
* maybe one podman Dockerfile in resources

And massive parts to be finished are :

* Documentation PR drafted
* Diskless PR

I propose we stop adding more roles for now (only the ones already in PR), and I will focus on fixing all issues and current PR.

**Comment by oxedions** [comment]
Once the 1.4 is ready, I propose before we release it to update the repositories organization, since it becomes a bit overcrowded.

![image](https://user-images.githubusercontent.com/15154435/99986936-1a14d900-2db0-11eb-9f0e-bab2ca732f40.png)

**bluebanquise** repository will be split in 4 repositories.

* **bluebanquise** will only contains core, adv_core, and add-ons that we, the core team, develop and support.
* **bluebanquise-community** will contains community add-ons. Each add-on will have a dedicated owner, the person who pushed it, or the person(s) who takes care of it. This ensures we (core team) can centralize community made add-ons around the stack, but are not directly in charge of them. <ins>Roles put here will still have to comply with BlueBanquise rules</ins>.
* **bluebanquise-disklesset** will contains the disklessset tool modules and main code.
* **bluebanquise-bootset** will contains the bootset tool modules and main code.

**bluebanquise-infrastructure** repository will be split in 2 repositories.

* **bluebanquise-infrastructure** will contains core packages needed files and the generator associated.
* **bluebanquise-infrastructure** will contains website files and the generator associated.

**bluebanquise-vagrant** remains unchanged.

All of this is made to ensure we can continue to provide best effort developments, debug, help, etc, without being flooded, as I see a huge wave coming soon...

**Reply by btravouillon** [reply]
Hi,

I do agree with this move, it will help to focus on the core and improve its quality while still providing some useful add-ons.
A few comments below.

### bluebanquise (core)

I would keep current directory names: core, advanced-core, addons
While writing, I wonder if we should move addons to core?

### bluebanquise-community

What would be the release model of **bluebanquise-community**? As a mono repo with different code owners per role, who will be responsible of the release management? 

Could we consider one repo per community role instead? Any role that comply with BlueBanquise rules could could be added to the bluebanquise namespace in GitHub.

### bluebanquise-{disklessset,bootset}

Both tools are written in python. Can you consider a repo **bluebanquise-tools** with both tools? It may be useful in the future when we will need to share some libraries between tools.

### bluebanquise-infrastructure

There is a typo in your post, **bluebanquise-website** is missing.

**Comment by strus38** [comment]
Just ensure that the update process is transparent, but it should given
what you said if the ansible.cfg is updated.

**Comment by strus38** [comment]
I fear that splitting too much will produce 'dead' repositories.

**Reply by btravouillon** [reply]
By 'dead' repositories, do you mean roles with no active maintainer?

**Comment by strus38** [comment]
Yes, and repositories which will not be updated / upgraded. I guess, it
will loose the advantage of being one stack... It will be a bunch of repos,
just like ansible Galaxy is.

**Reply by btravouillon** [reply]
The question is how a mono repo can solve this? Thus my question about the release model and release management.

If nobody care about a given role, should we keep this role in **bluebanquise-community** or drop it? Having a single repo or one repo per role does not solve the lack of maintainer for a given role.

**Comment by strus38** [comment]
At least when it is centralized, you can enforce the dev rules, inventory
rules,etc.
Concerning the maintenance, up to the bb team to decide what they want to
keep or drop. Anyone can then keep some roles separately, as horrible as he
wants... But it is no more your issue. Only the BB repo should be your
issue, and if it was only me, I would have put vagrant into it, and all
others repos which are for this project, to have a single repo.

**Reply by oxedions** [reply]
Thank you all for your feedbacks. 🙂 

I will compute all of this.

**Comment by oxedions** [comment, marked as answer]
We discussed all of this with @johnnykeats . Here is the summary:

**bluebanquise**

* Contains: core, adv_core
* Releases cycles: slow (4-6 month), with support on both dev (n+1) and latest version (n).
* Maintainers: core team.
* Owners: admins.
* Comments: key addons will be migrated to core, others will move to bluebanquise.community. List to be defined.

**bluebanquie.community**

* Contains: community addons
* Releases cycles: fast (~2 months), except if asked by a maintainer (urgent fix). Support only on dev version.
* Maintainers: each addon developer maintains it.
* Owners: each addon developer own it.
* Comments: 
    * if an addon is no more maintained, addon future is to be discussed by community (new maintainer, remove, etc).
    * releases done by core team.
    * main changelog to be updated at each PR modifying tasks (or related) or documentation. May migrate to changelog fragments if repository grow too much.

**bluebanquise.tools**

* Contains: tools (bootset, disklessset, etc.)
* Releases cycles: on need.
* Maintainers: core team.
* Owners: admins.

**bluebanquise-vragrant**

* Contains: vagrant auto deployment of BlueBanquise, to be used for core validation in future.
* Release cycle: on need.
* Maintainer: @btravouillon
* Owner: @btravouillon

**bluebanquise.website**

* Contains: bluebanquise.com website.
* Releases cycles: on need.
* Maintainers: core team.
* Owners: admins.

**bluebanquise.infrastructure**

* Contains: what is needed to build packages.
* Releases cycles: on need.
* Maintainers: core team for core packages, addon maintainer if addon packages.
* Owners: admins and community.


All repositories will be explicitly mentioned on the main BlueBanquise README.md, at top level.

Documentation "links" will be maintained, to include all readmes in the main documentation.

This structure should allow flexibility, and still provide visibility for all elements. It also provides owning for developers, which is something important for all.

Last point, @johnnykeats proposed, and I agree, that if a major feature/change is asked on a core role, and that core team do not wish to maintain or include such change, it is fully possible to fork a core role and create an addons with it in community repository. This to ensure/protect stability of the core while still allowing shared fork of it.

**Comment by oxedions** [comment]
1.4 is ready. Split of repositories will be done just after 1.4 release, so for 1.5.

**Reply by btravouillon** [reply]
FYI, before a release of 1.4, #445 introduced some bugs due to what looks like bad search & replace.

**Reply by oxedions** [reply]
Dear @btravouillon 

Thanks a lot ! Indeed, good catch 😊

**Reply by oxedions** [reply]
Solved

**Comment by oxedions** [comment]
@johnnykeats Could you take care of the new release ? Code name is "naquadah"
Under too much load 😪

**Reply by johnnykeats** [reply]
Ok. Can you build rpms?

**Reply by oxedions** [reply]
I could do the release 🙂
